src/data/zmod/basic.lean

```diff
@@ -577,6 +577,49 @@ def units_equiv_coprime {n : ℕ} [fact (0 < n)] :
   left_inv := λ ⟨_, _, _, _⟩, units.ext (nat_cast_zmod_val _),
   right_inv := λ ⟨_, _⟩, by simp }
 
+/-- The **Chinese remainder theorem**. For a pair of coprime natural numbers, `m` and `n`,
+  the rings `zmod (m * n)` and `zmod m × zmod n` are isomorphic. -/
+def chinese_remainder {m n : ℕ} (h : m.coprime n) :
+  zmod (m * n) ≃+* zmod m × zmod n :=
+if hmn0 : m * n = 0
+then if hm1 : m = 1
+  then by rw [hm1, one_mul]; exact ring_equiv.zero_ring_prod _ _
```

Comment:
I feel like using `rw` to construct data is going to make this reduce badly into terms with `eq.rec`. Can you use `fin.cast` or something instead?

Reply:
My feeling is that you shouldn't be reducing this particular definition anyway unless you have an explicit `m`, `n` and `x` in which case it will reduce just fine. It will be prohibitively ugly for anything else whatever I do.

Reply:
This has now been changed